### ISSUE TITLE
Pytorch implementation of DPGradientDescentGaussianOptimizer

### ISSUE THREAD

**Issue by srxzr** [tensorflow/privacy#42]
I implemented DPGradientDescentGaussianOptimizer in pytorch, however, I cannot get the similar accuracies as you mentioned in the tutorial. Using the default parameters in the current file I get about 94% accuracy I tried to tune the parameters I got about 95% accuracy at most . Do you have idea why this is happening?  

https://github.com/srxzr/DPSGD

**Comment by npapernot** [comment]
Just looking briefly at your code, it looks like your model has a larger fc layer. Could it be that the larger number of parameters interacts poorly with clipping?

**Comment by ebagdasa** [comment]
this is my implementation based on what I understood from the code. 

I think I am not 100% sure about the microbatches idea as the noise added the same if there are 256 microbatches or 32. 

```python
def train_dp(trainloader, model, optimizer, epoch, num_microbatches):
    model.train()
    running_loss = 0.0
    for i, data in tqdm(enumerate(trainloader, 0), leave=True):
        inputs, labels = data
        inputs = inputs.cuda()
        labels = labels.cuda()
        optimizer.zero_grad()

        outputs = model(inputs)
        loss = criterion(outputs, labels)
        running_loss += torch.mean(loss).item()

        losses = torch.mean(loss.reshape(num_microbatches, -1), dim=1)
        saved_var = dict()
        for tensor_name, tensor in model.named_parameters():
            saved_var[tensor_name] = torch.zeros_like(tensor)

        for j in losses:
            j.backward(retain_graph=True)
            torch.nn.utils.clip_grad_norm_(model.parameters(), S)
            for tensor_name, tensor in model.named_parameters():
                new_grad = tensor.grad
                saved_var[tensor_name].add_(new_grad)
            model.zero_grad()

        for tensor_name, tensor in model.named_parameters():
            saved_var[tensor_name].add_(torch.cuda.FloatTensor(tensor.grad.shape).normal_(0, sigma))
            tensor.grad = saved_var[tensor_name] / num_microbatches
        optimizer.step()
```

**Comment by ebagdasa** [comment]
ok, I posted the code here: https://github.com/ebagdasa/pytorch-privacy . Do I need to add more acknowledgements?  As well, I cut the corners and just implemented it using existing `clip_grad_norm_` method, but happy to add more ways.

**Comment by srxzr** [comment]
> Just looking briefly at your code, it looks like your model has a larger fc layer. Could it be that the larger number of parameters interacts poorly with clipping?

The number of trainable parameters in both model are similar. Actually the current code on my git has less parameters in in the fc layer compare to this code.

**Comment by srxzr** [comment]
> ok, I posted the code here: https://github.com/ebagdasa/pytorch-privacy . Do I need to add more acknowledgements? As well, I cut the corners and just implemented it using existing `clip_grad_norm_` method, but happy to add more ways.

did you get the same accuracy ? 
I ran the your code but with your parameters, the results are much lower than what I get .

**Comment by ebagdasa** [comment]
I get more than 98% with microbatches=256, it's just the default param I set was 32.

**Comment by srxzr** [comment]
> I get more than 98% with microbatches=256, it's just the default param I set was 32.

can you give me the best params ?

**Comment by ebagdasa** [comment]
```yaml
test_batch_size: 100
batch_size: 256
num_microbatches: 256
lr: 0.1
momentum: 0.5
decay: 0
S: 1
z: 1.1
dp: True
epochs: 60
save_on_epochs: [10, 50, 100, 200]
save_model: False
```

I get 95% after 10 epochs.

**Comment by srxzr** [comment]
> ```yaml
> test_batch_size: 100
> batch_size: 256
> num_microbatches: 256
> lr: 0.1
> momentum: 0.5
> decay: 0
> S: 1
> z: 1.1
> dp: True
> epochs: 60
> save_on_epochs: [10, 50, 100, 200]
> save_model: False
> ```
> 
> I get 95% after 10 epochs.

But  you said you can get up to 98% ? 
I ran your code with this parameters after 60 epoch I got about 96%

**Comment by ebagdasa** [comment]
you are right, I only could get 98% when I make noise multiplier lower, otherwise I am stuck with 96.5% at most and using lightweight model (the one from this repo and suggested by Nicolas) or stepping the learning rate doesn't help much. I will keep you posted.

**Comment by ebagdasa** [comment]
@srxzr,  did you get higher value (my best result is 96.7 with z=1.1)? Can we collaborate and get a reference PyTorch implementation that we can use later for the research.

**Comment by ebagdasa** [comment]
I think I am able to get 97.5 with lower lr (0.05), momentum 0.5 and training for 120 epochs.

**Comment by srxzr** [comment]
@ebagdasa but you are spending more privacy (120 epochs compare to 60).

**Comment by ebagdasa** [comment]
true, but it's still below 5. for 60 epochs I have 97.1. I think it's good enough given the network is different.

**Comment by schien1729** [comment]
Thank you for the conversation and updates! Closing this issue now as it isn't directly related to this repo.
